### PR TITLE
Change processId to be nullable so that users can specify 0 as processId

### PR DESCRIPTION
**Pull request by pieandcakes**
Allow processId to be 0 by changing it to be a nullable int.

### CODE REVIEW

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 1]
```diff
@@ -34,6 +34,8 @@ internal class AD7PendingBreakpoint : IDebugPendingBreakpoint2
         private bool _deleted;
         private bool _pendingDelete;
 
+        private uint hitCount;
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: >hitCount [](start = 21, length = 8)

Add `_` to follow the style.

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 2]
```diff
@@ -826,11 +826,11 @@ private void DetermineAndAddExecutablePathCommand(IList<LaunchCommand> commands,
             if (launchOptions.UnixPort.IsOSX())
             {
                 // Usually the first FD=txt in the output of lsof points to the executable.
-                absoluteExePath = string.Format(CultureInfo.InvariantCulture, "shell lsof -p {0} | awk '$4 == \"txt\" {{ print $9 }}'|awk 'NR==1 {{print $1}}'", _launchOptions.ProcessId);
+                absoluteExePath = string.Format(CultureInfo.InvariantCulture, "shell lsof -p {0} | awk '$4 == \"txt\" {{ print $9 }}'|awk 'NR==1 {{print $1}}'", _launchOptions.ProcessId.HasValue ? _launchOptions.ProcessId.Value : 0);
             }
             else if (launchOptions.UnixPort.IsLinux())
             {
-                absoluteExePath = string.Format(CultureInfo.InvariantCulture, @"shell readlink -f /proc/{0}/exe", _launchOptions.ProcessId);
+                absoluteExePath = string.Format(CultureInfo.InvariantCulture, @"shell readlink -f /proc/{0}/exe", _launchOptions.ProcessId.HasValue ? _launchOptions.ProcessId.Value : 0);
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: >HasValue [](start = 139, length = 8)

This method is only supposed to be called if the process id has a value. So I would suggest asserting and using `.Value`

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 3]
```diff
@@ -1173,7 +1173,7 @@ private void Dispose()
                     bool stoppedAtSIGSTOP = false;
                     if (sigName == "SIGSTOP")
                     {
-                        if (AD7Engine.RemoveChildProcess(_launchOptions.ProcessId))
+                        if (AD7Engine.RemoveChildProcess(_launchOptions.ProcessId.HasValue ? _launchOptions.ProcessId.Value : 0))
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: >RemoveChildProcess [](start = 38, length = 18)

I think you want to change the check on line 1175 to include `&& ProcessId.HasValue`.

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 4]
```diff
@@ -113,8 +115,9 @@ private bool CanBind()
             }
             if ((_bpRequestInfo.dwFields & enum_BPREQI_FIELDS.BPREQI_PASSCOUNT) != 0)
             {
-                this.SetError(new AD7ErrorBreakpoint(this, ResourceStrings.UnsupportedPassCountBreakpoint, enum_BP_ERROR_TYPE.BPET_GENERAL_ERROR));
-                return false;
+                hitCount = _bpRequestInfo.bpPassCount.dwPassCount;
+                //this.SetError(new AD7ErrorBreakpoint(this, ResourceStrings.UnsupportedPassCountBreakpoint, enum_BP_ERROR_TYPE.BPET_GENERAL_ERROR));
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: >SetError [](start = 23, length = 8)

If we are going to support hit count breakponits, don't we need to tell GDB about the hit count and reset the hit count in GDB if needed?

**Comment by pieandcakes** [comment]
@paulmaybee @gregg-miskelly sorry, I thought I had reset those before beginning my changes. I'll remove them

File: src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs [hunk 5]
```diff
@@ -114,7 +114,7 @@ private bool CanBind()
             if ((_bpRequestInfo.dwFields & enum_BPREQI_FIELDS.BPREQI_PASSCOUNT) != 0)
             {
                 this.SetError(new AD7ErrorBreakpoint(this, ResourceStrings.UnsupportedPassCountBreakpoint, enum_BP_ERROR_TYPE.BPET_GENERAL_ERROR));
-                return false;
+                return true;
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/AD7.Impl/AD7PendingBreakpoint.cs]: Shouldn't this part be backed out too?

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 6]
```diff
@@ -826,11 +826,12 @@ private void DetermineAndAddExecutablePathCommand(IList<LaunchCommand> commands,
             if (launchOptions.UnixPort.IsOSX())
             {
                 // Usually the first FD=txt in the output of lsof points to the executable.
-                absoluteExePath = string.Format(CultureInfo.InvariantCulture, "shell lsof -p {0} | awk '$4 == \"txt\" {{ print $9 }}'|awk 'NR==1 {{print $1}}'", _launchOptions.ProcessId);
+                absoluteExePath = string.Format(CultureInfo.InvariantCulture, "shell lsof -p {0} | awk '$4 == \"txt\" {{ print $9 }}'|awk 'NR==1 {{print $1}}'", _launchOptions.ProcessId.HasValue ? _launchOptions.ProcessId.Value : 0);
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: >ProcessId [](start = 176, length = 9)

This should use `.Value` also.

File: src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs [hunk 7]
```diff
@@ -826,11 +826,12 @@ private void DetermineAndAddExecutablePathCommand(IList<LaunchCommand> commands,
...
             }
             else if (launchOptions.UnixPort.IsLinux())
             {
-                absoluteExePath = string.Format(CultureInfo.InvariantCulture, @"shell readlink -f /proc/{0}/exe", _launchOptions.ProcessId);
+                Debug.Assert(_launchOptions.ProcessId.HasValue, "ProcessId should have a value.");
```
Comment (gregg-miskelly) [inline src/MIDebugEngine/Engine.Impl/DebuggedProcess.cs]: >Debug [](start = 16, length = 5)

Move the assert to the top of the function.

**Comment by gregg-miskelly** [comment]
Otherwise LGTM

**Pull request merged**